content/posts/type_safe_expressjs_genezio.md

```diff
@@ -0,0 +1,270 @@
+
+```bash
+genezio sdk --tarball --source local --packageName genezio-sdk  --output ../client-ts/sdk --language ts --url http://localhost:8881/genezio
+```
+
+If you're curious, take a look at what was generated. It’s a tarball NPM package, which means it can be installed locally just like any other NPM package. It could also be published to a public registry such as the NPM Registry, or to a private registry. You can also export the SDK as raw files by omitting the --tarball option. This allows you to use npm link if you wish, or to add a README, or modify the package.json before publishing it.
+
+For the sake of this tutorial, we are going to install the SDK locally. In `./client-ts` folder, we can run `npm install ./sdk/genezio-sdk.tar.gz` which will install the SDK in your `node_modules` folder. 
+
+```bash
+npm install ./sdk/genezio-sdk.tar.gz
+```
+
+We are now ready to change the communication between React and the server to use genezio. Go to `src/views.tsx` and import the `BookService` and the `Book` model.
+
+```ts
+import { BookService, Book } from "genezio-sdk"
+```
+
+You can remove the `requests.ts` file from the project and you can uninstall `axios`. It feels great when you remove code, doesn't it? Now we just have to use `BookService.getAllBooks` and `BookService.createBook` from the genezio SDK to call our backend methods.
+
+```ts
+  // Fetch all books when the component mounts
+  useEffect(() => {
+    const loadBooks = async () => {
+      const allBooks = await BookService.getAllBooks();
+      setBooks(allBooks);
+    };
+
+    loadBooks();
+  }, []);
+
+  const handleCreateBook = async (e: any) => {
+    e.preventDefault(); // Prevent form submission from reloading the page
+
+    const newBook = await BookService.createBook(title, author);
+    setBooks([...books, newBook]); // Add the new book to the local state
+
+    // Reset the form fields
+    setTitle('');
+    setAuthor('');
+  };
+```
+
+What is great now is that we can leverage the IDE capabilities. Hover over the `BookService.getAllBooks()` method and you can see: the documentation, the parameter types and the return type of the method. Pass a wrong parameter to `BookService.createBook()` and your IDE will shout that there is something wrong.
+
+![Alt text](/posts/demo-express-genezio.gif)
+
+Let's see one more thing: how do we handle errors? Easy! Just like you would normally do with Promises, you can either call `.catch` or wrap the `await` call around a `try catch`.
+
+```ts
+try {
+    const newBook = await BookService.createBook(title, author);
+    setBooks([...books, newBook]); // Add the new book to the local state
+
+    // Reset the form fields
+    setTitle('');
+    setAuthor('');
+} catch (error) {
+    console.error('Failed to create book:', error);
+    alert('Failed to create book');
+}
+```
+
```

Comment:
Here I would have liked to have some instruction on how to test the application locally:

In the `server` I should run `npm run build && npm start`.
In the `client-ts`, I should run `npm run dev`.